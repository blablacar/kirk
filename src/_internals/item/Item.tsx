import React from 'react'
import cc from 'classcat'

import { color } from '../../_utils/branding'
import { A11yProps, pickA11yProps } from '../../_utils/interfaces'
import { Button } from '../../button/Button'
import { ChevronIcon } from '../../icon/chevronIcon'
import { NormalizeProps } from '../../layout/layoutNormalizer'
import { Text, TextDisplayType, TextTagType } from '../../text'

export enum ItemStatus {
  DEFAULT = 'default',
  LOADING = 'loading',
  CHECKED = 'checked',
}

export type ItemProps = A11yProps &
  NormalizeProps &
  Readonly<{
    chevron?: boolean
    className?: string
    href?: string | JSX.Element
    highlighted?: boolean
    isClickable?: boolean
    isWrappable?: boolean
    leftTitle?: React.ReactNode
    leftTitleButtonAddon?: React.ReactElement<Button>
    leftTitleDisplay?: TextDisplayType
    leftTitleColor?: string
    leftBody?: string | React.ReactNode
    leftBodyDisplay?: TextDisplayType
    leftBodyColor?: string
    leftBodyAnnotation?: React.ReactNode
    leftBodyAnnotationDisplay?: TextDisplayType
    leftBodyAnnotationColor?: string
    leftAddon?: React.ReactNode
    rightTitle?: string | JSX.Element
    rightTitleDisplay?: TextDisplayType
    rightTitleStrikeThrough?: boolean
    rightTitleAriaProps?: A11yProps
    rightTitleColor?: string
    rightBody?: React.ReactNode
    rightBodyDisplay?: TextDisplayType
    rightBodyColor?: string
    rightAddon?: React.ReactNode
    tag?: JSX.Element
    onClick?: (event: React.MouseEvent<HTMLElement>) => void
    onBlur?: (event: React.FocusEventHandler<HTMLElement>) => void
    onFocus?: (event: React.FocusEventHandler<HTMLElement>) => void
    onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
    hideHoverBackground?: boolean
    disabled?: boolean
  }>

export const Item = (props: ItemProps) => {
  const {
    chevron,
    className,
    href,
    onClick,
    onBlur,
    onFocus,
    onMouseDown,
    highlighted,
    isClickable,
    isWrappable,
    leftTitle,
    leftTitleButtonAddon,
    rightTitleStrikeThrough,
    rightTitleAriaProps,
    leftTitleDisplay = TextDisplayType.TITLE,
    leftTitleColor,
    leftBody,
    leftBodyDisplay = TextDisplayType.BODY,
    leftBodyColor,
    leftBodyAnnotation,
    leftBodyAnnotationDisplay,
    leftBodyAnnotationColor,
    leftAddon,
    rightTitle,
    rightTitleDisplay = TextDisplayType.TITLE,
    rightTitleColor,
    rightBody,
    rightBodyDisplay = TextDisplayType.BODY,
    rightBodyColor,
    rightAddon,
    tag = <div />,
    hideHoverBackground = false,
    disabled = false,
  } = props
  const a11yAttrs = pickA11yProps<ItemProps>(props)
  let Tag = tag.type
  let tagProps = tag.props

  if (href) {
    if (typeof href !== 'string') {
      Tag = href.type
      tagProps = { ...tagProps, ...href.props }
    } else {
      Tag = 'a'
      tagProps = { href }
    }
  }

  if (Tag === 'button') {
    // The code below make sure that if we use a HTML button, a proper type='button' or
    // type='submit' is added. type='button' is the default.
    // This type=button is needed to make sure non-submit buttons are not activated when
    // submitting forms.
    // See: https://stackoverflow.com/questions/4763638/enter-triggers-button-click
    const buttonType = tag.props.type || 'button'
    tagProps = { ...tagProps, type: buttonType }
  }

  const hasRightText = rightTitle || rightBody

  const getTextColor = (textColor: string) => (disabled ? color.gray : textColor)

  const getLeftTitle = (value: React.ReactNode): React.ReactNode => {
    if (React.isValidElement(value)) {
      return leftTitle
    }
    return (
      <Text
        className={leftTitleButtonAddon ? 'kirk-item-title--withButtonAddon' : null}
        display={leftTitleDisplay}
        textColor={getTextColor(leftTitleColor)}
        tag={TextTagType.SPAN}
      >
        {leftTitle}
        {/* In case of a clickable Item don't display the addon button */}
        {!href && leftTitleButtonAddon}
      </Text>
    )
  }

  return (
    <Tag
      {...tagProps}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      onMouseDown={onMouseDown}
      className={cc([
        'kirk-item',
        {
          'kirk-item--highlighted': highlighted,
          'kirk-item--clickable': isClickable,
          'kirk-item--wrappable': isWrappable,
          'kirk-item--hideHoverBackground': hideHoverBackground,
        },
        className,
      ])}
      {...a11yAttrs}
    >
      <span className="kirk-item-leftWrapper">
        {leftAddon && <span className="kirk-item-leftAddon">{leftAddon}</span>}
        <span className="kirk-item-leftText">
          {leftTitle && getLeftTitle(leftTitle)}
          {leftBody && (
            <Text
              className="kirk-item-body"
              display={leftBodyDisplay}
              textColor={getTextColor(leftBodyColor)}
              tag={TextTagType.SPAN}
            >
              {leftBody}
            </Text>
          )}
          {leftBodyAnnotation && (
            <Text
              className="kirk-item-body-annotation"
              display={leftBodyAnnotationDisplay}
              textColor={getTextColor(leftBodyAnnotationColor)}
              tag={TextTagType.SPAN}
            >
              {leftBodyAnnotation}
            </Text>
          )}
        </span>
      </span>
      {hasRightText && (
        <span className="kirk-item-rightText">
          {rightTitle && (
            <Text
              className={cc([
                'kirk-item-title',
                {
                  'kirk-item--strikethrough': rightTitleStrikeThrough,
                },
              ])}
              display={rightTitleDisplay}
              textColor={getTextColor(rightTitleColor)}
              tag={TextTagType.DIV}
              {...rightTitleAriaProps}
            >
              {rightTitle}
            </Text>
          )}
          {rightBody && (
            <Text
              className="kirk-item-body"
              display={rightBodyDisplay}
              textColor={getTextColor(rightBodyColor)}
              tag={TextTagType.DIV}
            >
              {rightBody}
            </Text>
          )}
        </span>
      )}
      {rightAddon && <span className="kirk-item-rightAddon">{rightAddon}</span>}
      {chevron && (
        <span className="kirk-item-rightAddon">
          <ChevronIcon iconColor={!isClickable ? color.gray : color.lightMidnightGreen} />
        </span>
      )}
    </Tag>
  )
}
