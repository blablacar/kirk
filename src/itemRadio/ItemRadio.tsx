import React, { Component, Fragment } from 'react'
import cc from 'classcat'

import { OnChangeParameters } from '../_internals/onChange'
import { RadioIcon } from '../_internals/radioIcon'
import { A11yProps, pickA11yProps } from '../_utils/interfaces'
import { NormalizeProps } from '../layout/layoutNormalizer'
import { TextDisplayType } from '../text'
import { StyledItemRadio } from './ItemRadio.style'

export enum ItemRadioStatus {
  DEFAULT = 'default',
  LOADING = 'loading',
}

export type ItemRadioProps = A11yProps &
  NormalizeProps &
  Readonly<{
    className?: string
    name: string
    value: string | number
    leftAddon?: React.ReactNode
    labelTitle?: string
    label: string
    data?: string
    dataInfo?: string
    checked?: boolean
    disabled?: boolean
    chevron?: boolean
    highlighted?: boolean
    onChange?: (obj: OnChangeParameters) => void
    onClick?: (obj: OnChangeParameters) => void
    status?: ItemRadioStatus
    key?: string | number
  }>

type ItemRadioState = {
  focus: boolean
}

export class ItemRadio extends Component<ItemRadioProps> {
  static defaultProps: Partial<ItemRadioProps> = {
    onChange() {},
    onClick() {},
    checked: false,
  }

  state: ItemRadioState = {
    focus: false,
  }

  onChange = () => {
    const { name, value } = this.props
    this.props.onChange({ name, value })
  }

  onClick = () => {
    const { name, value } = this.props
    this.props.onClick({ name, value })
  }

  onFocus = () => {
    this.setState({ focus: true })
  }

  onBlur = () => {
    this.setState({ focus: false })
  }

  render() {
    const {
      label,
      name,
      value,
      data,
      className,
      labelTitle,
      dataInfo,
      checked,
      disabled,
      status,
      chevron,
      highlighted,
      leftAddon,
      hasHorizontalSpacing = false,
    } = this.props
    const a11yAttrs = pickA11yProps<ItemRadioProps>(this.props)
    const isLoading = status === ItemRadioStatus.LOADING
    const radio = (
      <Fragment>
        <input
          type="radio"
          name={name}
          value={value}
          checked={checked}
          onChange={this.onChange}
          onClick={this.onClick}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          disabled={disabled || isLoading}
        />
        {(!chevron || isLoading) && (
          <RadioIcon isChecked={checked} isLoading={isLoading} isDisabled={disabled} />
        )}
      </Fragment>
    )

    return (
      <StyledItemRadio
        className={cc(['kirk-item-radio', className, { focus: this.state.focus }])}
        leftTitle={labelTitle}
        leftBody={label}
        rightTitle={data}
        rightTitleDisplay={TextDisplayType.SUBHEADERSTRONG}
        rightBody={dataInfo}
        /* No a11y issue here
            - The input is well wrapped with the label
            - The linter can't access the complex components implementation
          */
        // eslint-disable-next-line jsx-a11y/label-has-associated-control
        tag={<label />}
        leftAddon={leftAddon}
        rightAddon={radio}
        chevron={chevron && !isLoading}
        highlighted={highlighted}
        isClickable={!disabled}
        disabled={disabled}
        hasHorizontalSpacing={hasHorizontalSpacing}
        {...a11yAttrs}
      />
    )
  }
}
