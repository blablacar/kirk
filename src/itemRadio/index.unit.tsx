import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import Item from '_utils/item'
import ItemRadio, { ItemRadioProps } from './index'

describe('ItemRadio', () => {
  const defaultProps: ItemRadioProps = {
    label: 'Label',
    name: 'name',
    value: 0,
    className: 'custom-class-name',
    labelTitle: 'Label title',
    data: 'Data',
    dataInfo: 'Data info',
    checked: false,
    disabled: false,
    onChange() {},
    status: ItemRadio.STATUS.DEFAULT,
    key: 0,
  }

  it('Should use the Item component', () => {
    const itemRadio = shallow(<ItemRadio {...defaultProps} />)
    expect(itemRadio.find(Item).exists()).toBe(true)
  })
  it('Should forward its props to the Item component', () => {
    const itemRadio = renderer.create(<ItemRadio {...defaultProps} />).toJSON()
    expect(itemRadio).toMatchSnapshot()
  })
  it('Should display a CircleIcon with an innerDisc when the input is checked', () => {
    const itemRadio = renderer.create(<ItemRadio {...defaultProps} checked />).toJSON()
    expect(itemRadio).toMatchSnapshot()
  })
  it('Should display a Loader when the component is in loading status', () => {
    const itemRadio = renderer
      .create(<ItemRadio {...defaultProps} status={ItemRadio.STATUS.LOADING} />)
      .toJSON()
    expect(itemRadio).toMatchSnapshot()
  })
  it('Should bind the onChange callback of the input', () => {
    const itemRadio = mount(<ItemRadio {...defaultProps} />)
    expect(itemRadio.find('input').prop('onChange')).toEqual(itemRadio.instance().onChange)
  })
  it('Should call the onChange prop with name and value when the input changes', () => {
    const onChangeMock = jest.fn()
    const itemRadio = shallow(<ItemRadio {...defaultProps} onChange={onChangeMock} />)
    itemRadio.instance().onChange()
    expect(onChangeMock).toHaveBeenCalledWith({
      name: defaultProps.name,
      value: defaultProps.value,
    })
  })
})
