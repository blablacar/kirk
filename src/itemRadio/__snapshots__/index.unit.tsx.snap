// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemRadio Should display a CircleIcon with an innerDisc when the input is checked 1`] = `
<label
  className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-radio custom-class-name"
>
  <div
    className="jsx-1542968643 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightAddon"
  >
    <input
      checked={true}
      disabled={false}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <svg
      aria-hidden={true}
      className="kirk-icon kirk-icon-circle"
      fill="#00AFF5"
      height={24}
      viewBox="0 0 66 66"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <circle
        className="jsx-2931153156"
        cx="33"
        cy="33"
        fill="none"
        r="30"
        stroke="#00AFF5"
      />
      <circle
        className="jsx-2931153156 inner"
        cx="33"
        cy="33"
        fill="#00AFF5"
        r="18"
      />
    </svg>
  </div>
</label>
`;

exports[`ItemRadio Should display a Loader when the component is in loading status 1`] = `
<label
  className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-radio custom-class-name"
>
  <div
    className="jsx-1542968643 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightAddon"
  >
    <input
      checked={false}
      disabled={true}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <div
      className="jsx-2765694959 kirk-loader"
      style={
        Object {
          "height": "24px",
          "width": "24px",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="kirk-icon kirk-icon-circle spinning"
        fill="#5DD167"
        height={24}
        viewBox="0 0 66 66"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <circle
          className="jsx-2931153156"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          stroke="#5DD167"
        />
      </svg>
    </div>
  </div>
</label>
`;

exports[`ItemRadio Should forward its props to the Item component 1`] = `
<label
  className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-radio custom-class-name"
>
  <div
    className="jsx-1542968643 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1542968643 kirk-item-rightAddon"
  >
    <input
      checked={false}
      disabled={false}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <svg
      aria-hidden={true}
      className="kirk-icon kirk-icon-circle"
      fill="#00AFF5"
      height={24}
      viewBox="0 0 66 66"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <circle
        className="jsx-2931153156"
        cx="33"
        cy="33"
        fill="none"
        r="30"
        stroke="#00AFF5"
      />
    </svg>
  </div>
</label>
`;
